src/script/tari_script.rs

```diff
@@ -428,6 +467,42 @@ impl TariScript {
             (..) => Err(ScriptError::IncompatibleTypes),
         }
     }
+
+    fn check_multisig(
+        &self,
+        stack: &mut ExecutionStack,
+        m: u8,
+        n: u8,
+        public_keys: &Vec<RistrettoPublicKey>,
+        message: Message,
+    ) -> Result<bool, ScriptError> {
+        if m == 0 || n == 0 || m > n || n > MAX_MULTISIG_LIMIT {
+            return Err(ScriptError::InvalidData);
+        }
+        // pop m sigs
+        let signatures = stack
+            .pop_num_items(m.into())?
+            .into_iter()
+            .map(|item| match item {
+                StackItem::Signature(s) => Ok(s),
+                _ => Err(ScriptError::IncompatibleTypes),
+            })
+            .collect::<Result<Vec<RistrettoSchnorr>, ScriptError>>()?;
+
+        let mut valid_signatures = 0;
+        let mut key_signed = vec![false; public_keys.len()];
+        for (i, public_key) in public_keys.iter().enumerate() {
+            for signature in signatures.iter() {
+                if !key_signed[i] && signature.verify_challenge(public_key, &message) {
+                    valid_signatures += 1;
+                    key_signed[i] = true;
+                    break;
+                }
+            }
+        }
```

Comment:
This catches the issue where one key submits 2 sigs, but I don't think it solves the case @mikethetike raised?

Best to add a test for submitting 3 of 5 multisig where pubkey 1 and 3 are repeated, and sig 3 is signed by that key.

Reply:
Should the supplied public keys be enforced to be unique?

Reply:
We can allow duped pubkeys, but we should enforce that each signature is used on at most one pubkey, and that each pubkey is used at most once.

A simpler and more direct test would be
MultiSig(2, 3, [P_a, P_b, P_a]) with [s_a, s_x] should fail (s_a cannot be used on both P_as to get 2 valid signatures)

Reply:
Just added that fix and a test for it

Reply:
updated the test to add your test case: [link](https://github.com/tari-project/tari-crypto/pull/59/commits/31a1e5d4a4456bee9e1d9ee04f32b2ff51eabd5c#diff-2b1d02202ea28e771ea738d8c773f3dee0f3d216c5afdd6c613781709f97b33eR1071)